index.d.ts

```diff
@@ -0,0 +1,60 @@
+export interface Action {
+  type: string;
+}
+
+export type Reducer<S> = (state: S, action: Action) => S;
+
+export type Dispatch = (action: any) => any;
+
+export interface MiddlewareAPI<S> {
+  dispatch: Dispatch;
+  getState: () => S;
+}
+
+export interface Middleware {
+  <S>(api: MiddlewareAPI<S>): (next: Dispatch) => Dispatch;
+}
```

Comment:
Let's try to implement `thunk` middleware in TypeScript as an example. Start without any types:
```ts
const thunkMiddleware = ({dispatch, getState}) =>
  (next) => (action) => {
    return typeof action === function ? action(dispatch, getState) : next(action)
  }
```
Now what types can we add here? Keep in mind that there may me other middlewares applied before `thunk`, so `dispatch` here can potentially accept anything, e.g. promises. Same for `next`, same for `action`.
Middleware is standalone and doesn't know anything about `dispatch` type prior to when it was applied.